crmsh/bootstrap.py

```diff
@@ -1692,6 +1693,106 @@ def init():
     init_network()
 
 
+class JoinLock(object):
+    """
+    Class to manage lock for multiple nodes join in parallel
+    """
+
+    JOIN_LOCK_FILE = "/usr/share/crmsh/.join_lock_file"
+    DETECT_CMD = "test -f {}".format(JOIN_LOCK_FILE)
+    TAIL_CMD = "tail -1 {}".format(JOIN_LOCK_FILE)
+    RM_CMD = "rm -f {}".format(JOIN_LOCK_FILE)
+    TOUCH_CMD = "touch {}".format(JOIN_LOCK_FILE)
+    UPDATE_CMD = "echo {} `date +%s` >> {}".format(utils.this_node(), JOIN_LOCK_FILE)
+
+    def __init__(self, target_node):
+        """
+        Init function
+        """
+        self.thread_stop_event = None
+        self.target_node = target_node
+
+    def _run(self, cmd):
+        """
+        Wrapper utils.run_cmd_on_remote, exit when meeting error
+        """
+        rc, out, err = utils.run_cmd_on_remote(cmd, self.target_node)
+        if rc != 0 and err:
+            error(err)
+        return rc, out
+
+    def _detect(self):
+        """
+        Detect whether target node has the lock file
+        """
+        rc, _ = self._run(self.DETECT_CMD)
+        return rc == 0
+
+    def _wait(self):
+        """
+        Wait 15 seconds while detect target node has lock file
+        """
+        warned_once = False
+        origin_timestamp = None
+
+        while self._detect():
+            # tail the last line in lock file
+            rc, out = self._run(self.TAIL_CMD)
```

Comment:
Maybe we could create methods for all of these `self._run` methods, like `_get_tail`.
To make it a bit more meaningful. The same you do with `_detect`.
This way, for example with `TAIL_CMD`, we could put the all of this logic inside the method, and only return the timestamp as `int`
It is not wrong anyway

Reply:
OK, changed, create some meaningful and small methods, and put part of logic into `_get_last_timestamp`